### ISSUE TITLE
Optimization error with pyOptSparseDriver with linear constraints

### ISSUE THREAD

**Issue by tuckerbabcock** [OpenMDAO/OpenMDAO#2647]
### Description

When using linear constraints with pyOptSparse, I've found that the optimization does not behave as I'd expect it to, and have been able to create the MWE that demonstrates some of the behavior I've seen. When using the linear constraint, the optimization "converges", with pyOptSparse thinking it has satisfied the constraint, when the constraint is actually violated. Running the example I've included as is I see this output:
```
Optimization Problem -- Optimization using pyOpt_sparse
================================================================================
    Objective Function: _objfunc

    Solution: 
--------------------------------------------------------------------------------
    Total Time:                    0.0033
       User Objective Time :       0.0012
       User Sensitivity Time :     0.0007
       Interface Time :            0.0010
       Opt Solver Time:            0.0004
    Calls to Objective Function :       5
    Calls to Sens Function :            4


   Objectives
      Index  Name                  Value          Optimum
          0  parab.f_xy     1.300000E+01     0.000000E+00

   Variables (c - continuous, i - integer, d - discrete)
      Index  Name   Type      Lower Bound            Value      Upper Bound     Status
          0  y_0      c    -5.000000E+01     0.000000E+00     5.000000E+01           

   Constraints (i - inequality, e - equality)
      Index  Name    Type          Lower           Value           Upper    Status  Lagrange Multiplier
          0  const.g    i   0.000000E+00    0.000000E+00    1.000000E+30         l     1.10000E+01

--------------------------------------------------------------------------------

Driver debug print for iter coord: rank0:pyOptSparse_SNOPT|6
------------------------------------------------------------
Design Vars
{'y': array([0.])}

Nonlinear constraints
None

Linear constraints
{'const.g': array([3.])}

Objectives
{'parab.f_xy': array([13.])}
```
The summary prints that the constraint value is `0.0`,  but the debug printing shows that the linear constraint has a value of `3.0`.

If I switch the constraint the be nonlinear (by uncommenting `prob.model.add_constraint('const.g', lower=0., linear=False)` and commenting `prob.model.add_constraint('const.g', lower=0., linear=True)`), this is the output I see (and expect):
```
Optimization Problem -- Optimization using pyOpt_sparse
================================================================================
    Objective Function: _objfunc

    Solution: 
--------------------------------------------------------------------------------
    Total Time:                    0.0048
       User Objective Time :       0.0009
       User Sensitivity Time :     0.0030
       Interface Time :            0.0005
       Opt Solver Time:            0.0004
    Calls to Objective Function :       3
    Calls to Sens Function :            2


   Objectives
      Index  Name                  Value          Optimum
          0  parab.f_xy    -1.100000E+01     0.000000E+00

   Variables (c - continuous, i - integer, d - discrete)
      Index  Name   Type      Lower Bound            Value      Upper Bound     Status
          0  y_0      c    -5.000000E+01    -3.000000E+00     5.000000E+01           

   Constraints (i - inequality, e - equality)
      Index  Name    Type          Lower           Value           Upper    Status  Lagrange Multiplier
          0  const.g    i   0.000000E+00    0.000000E+00    1.000000E+30         l     5.00000E+00

--------------------------------------------------------------------------------

Driver debug print for iter coord: rank0:pyOptSparse_SNOPT|3
------------------------------------------------------------
Design Vars
{'y': array([-3.])}

Nonlinear constraints
{'const.g': array([0.])}

Linear constraints
None

Objectives
{'parab.f_xy': array([-11.])}
```

At first I thought that the `ExecComp` might somehow be the issue, but behavior is the same no matter if I use that or use the `Constraint` class. If I replacing the `pyOptSparseDriver` with the `ScipyOptimizeDriver`, the optimization works correctly for both the linear and the nonlinear cases.

### Example

```python
import openmdao.api as om
from openmdao.test_suite.components.paraboloid import Paraboloid

class Constraint(om.ExplicitComponent):
    def setup(self):
        self.add_input('x', val=0.0)
        self.add_input('y', val=0.0)

        self.add_output('g', val=0.0)

    def setup_partials(self):
        self.declare_partials('*', '*')

    def compute(self, inputs, outputs):
        x = inputs['x']
        y = inputs['y']

        outputs['g'] = x + y

    def compute_partials(self, inputs, partials):
        partials['g', 'x'] = 1.0
        partials['g', 'y'] = 1.0

# build the model
prob = om.Problem()
prob.model.add_subsystem('parab', Paraboloid(), promotes_inputs=['x', 'y'])

### The problem exists with the ExecComp version of the constraint or the
### custom Constraint class
# define the component whose output will be constrained
# prob.model.add_subsystem('const', om.ExecComp('g = x + y'), promotes_inputs=['x', 'y'])
prob.model.add_subsystem('const', Constraint(), promotes_inputs=['x', 'y'])

# Design variables 'x' and 'y' span components, so we need to provide a common initial
# value for them.
prob.model.set_input_defaults('x', 3.0)
prob.model.set_input_defaults('y', -4.0)

# Set up optimization driver
prob.driver = om.pyOptSparseDriver()
prob.driver.options['optimizer'] = 'SNOPT'
prob.driver.options['debug_print'] = ['desvars', 'ln_cons', 'nl_cons', 'objs', 'totals']

### Things seem to work correctly with the ScipyOptimizeDriver
# prob.driver = om.ScipyOptimizeDriver()
# prob.driver.options['optimizer'] = 'SLSQP'

# prob.model.add_design_var('x', lower=-50, upper=50)
prob.model.add_design_var('y', lower=-50, upper=50)
prob.model.add_objective('parab.f_xy')

### Switch the commenting of these to quickly switch the constraint type
# prob.model.add_constraint('const.g', lower=0., linear=False)
prob.model.add_constraint('const.g', lower=0., linear=True)

prob.setup()
prob.run_driver();
```

### OpenMDAO Version

3.20.0

### Relevant environment information

`pip freeze`:
```
brotlipy==0.7.0
certifi==2021.10.8
cffi @ file:///Users/runner/miniforge3/conda-bld/cffi_1631636293358/work
chardet @ file:///Users/runner/miniforge3/conda-bld/chardet_1610093454858/work
charset-normalizer @ file:///home/conda/feedstock_root/build_artifacts/charset-normalizer_1626371162869/work
clingo==5.5.1
colorama @ file:///home/conda/feedstock_root/build_artifacts/colorama_1602866480661/work
conda==4.11.0
conda-package-handling @ file:///Users/runner/miniforge3/conda-bld/conda-package-handling_1618231726924/work
coverage==4.5.4
cryptography @ file:///Users/runner/miniforge3/conda-bld/cryptography_1633983374178/work
cycler @ file:///home/conda/feedstock_root/build_artifacts/cycler_1635519461629/work
Cython @ file:///Users/runner/miniforge3/conda-bld/cython_1645125497141/work
fortls==2.6.0
idna @ file:///home/conda/feedstock_root/build_artifacts/idna_1609836280497/work
-e git+ssh://git@github.com/mdolab/idwarp.git@a1f6b31f491881f31bc0b0c747b47b7e2373c191#egg=idwarp
kiwisolver @ file:///Users/runner/miniforge3/conda-bld/kiwisolver_1635836816114/work
-e git+ssh://git@github.com/OptimalDesignLab/mach.git@f2b27271b44ac61f44ff3abfca937523f68ce8c8#egg=mach
matplotlib @ file:///Users/runner/miniforge3/conda-bld/matplotlib-suite_1632416631285/work
mdolab-baseclasses==1.6.1
-e git+ssh://git@github.com/tuckerbabcock/MotorModel.git@6486164c6fd44a1927a4c1c906e8b245f0bf9296#egg=motormodel
-e git+ssh://git@github.com/tuckerbabcock/mphys.git@c67212caca8ecb4a934b4b4bb05c1989b7041f03#egg=mphys
mpi4py==3.1.3
networkx @ file:///home/conda/feedstock_root/build_artifacts/networkx_1635253012265/work
numpy @ file:///Users/runner/miniforge3/conda-bld/numpy_1636145451503/work
olefile @ file:///home/conda/feedstock_root/build_artifacts/olefile_1602866521163/work
-e git+ssh://git@github.com/tuckerbabcock/omESP.git@71c4c26fe830c3946246ed4cd53733b597af8919#egg=omESP
openmdao==3.20.0
packaging==21.3
pandas==1.3.4
parameterized==0.8.1
petsc4py==3.17.2
Pillow @ file:///Users/runner/miniforge3/conda-bld/pillow_1630696664467/work
pycosat @ file:///Users/runner/miniforge3/conda-bld/pycosat_1624996792650/work
pycparser @ file:///home/conda/feedstock_root/build_artifacts/pycparser_1593275161868/work
pyDOE2==1.3.0
pyOpenSSL @ file:///home/conda/feedstock_root/build_artifacts/pyopenssl_1633192417276/work
pyoptsparse @ file:///Users/tuckerbabcock/Developer/motor/build_pyoptsparse/build_pyoptsparse.628e57dd/pyoptsparse
pyparsing==3.0.9
PySocks @ file:///Users/runner/miniforge3/conda-bld/pysocks_1610291460269/work
python-dateutil @ file:///home/conda/feedstock_root/build_artifacts/python-dateutil_1626286286081/work
pytz @ file:///home/conda/feedstock_root/build_artifacts/pytz_1633452062248/work
pyXDSM==2.2.1
-e git+ssh://git@github.com/OpenMDAO/RadialMotor.git@d2951775ea58c77211af950576fb58c73c5bba52#egg=rad_motor
requests @ file:///home/conda/feedstock_root/build_artifacts/requests_1626393743643/work
ruamel-yaml-conda @ file:///Users/runner/miniforge3/conda-bld/ruamel_yaml_1624999868738/work
scipy @ file:///Users/runner/miniforge3/conda-bld/scipy_1633644994244/work
six @ file:///home/conda/feedstock_root/build_artifacts/six_1620240208055/work
sqlitedict==2.0.0
testflo==1.4.8
tornado @ file:///Users/runner/miniforge3/conda-bld/tornado_1635819711289/work
tqdm @ file:///home/conda/feedstock_root/build_artifacts/tqdm_1632160078689/work
urllib3 @ file:///home/conda/feedstock_root/build_artifacts/urllib3_1632350318291/work
```

**Comment by Kenneth-T-Moore** [comment]
This bug is fixed in #2605 .  It was a problem with how pyoptsparse handles linear constraints, assuming that the y-intercept is 0. You can use the latest fix by cloning the openmdao repository with the current master.  The fix will also be on the next release, which should be pretty soon, as the last release was on August 10.

**Comment by tuckerbabcock** [comment]
Thank you @Kenneth-T-Moore!
